### DISCUSSION TITLE
Is there a list of what async primitive MUST be used with a tokio runtime?

### DISCUSSION THREAD

**Discussion by tisonkun** [tokio-rs/tokio#6502] [Q&A] [answered]
Most of `Future`s can be polled with any runtime, but some of the primitives, like `tokio::net::TcpListener`, `tokio::sync::oneshot` must be used with tokio runtime since they cooperate with the TLS handle.

Is there a list of what async primitive MUST be used with a tokio runtime?

**Comment by Darksonn** [comment, marked as answer]
The things in `tokio::sync` do not need a runtime, and everything else needs a Tokio runtime.

You mention the oneshot channel, but it does not depend on the runtime. Yes, it will access the Tokio's TLS storage to interact with [cooperative scheduling](https://docs.rs/tokio/latest/tokio/task/index.html#cooperative-scheduling), but this is optional, and the channel will still work if the TLS is unavailable.

**Reply by tisonkun** [reply]
> The things in tokio::sync do not need a runtime, and everything else needs a Tokio runtime.

This is the answer I want to know. And good to know for the details of oneshot channel. Thank you!
